src/coreclr/vm/comdelegate.cpp

```diff
@@ -801,49 +913,59 @@ static Stub* SetupShuffleThunk(MethodTable* pDelMT, MethodDesc* pTargetMeth)
+    // hidden return buffer argument because the extra argument switches place with the target when coming from the caller.
+    Stub* pShuffleThunk = isInstRetBuff ? pClass->m_pInstRetBuffCallStub : pClass->m_pStaticCallStub;
+    if (pShuffleThunk)
+        return pShuffleThunk->GetEntryPoint();
+
+    GCX_PREEMP();
+
     MethodDesc *pMD = pClass->GetInvokeMethod();
 
+    // We haven't already setup a shuffle thunk, go do it now (which will cache the result automatically).
     StackSArray<ShuffleEntry> rShuffleEntryArray;
-    GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray);
+    if (GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray))
+    {
+        ShuffleThunkCache* pShuffleThunkCache = s_pShuffleThunkCache;
 
-    ShuffleThunkCache* pShuffleThunkCache = s_pShuffleThunkCache;
+        LoaderAllocator* pLoaderAllocator = pDelMT->GetLoaderAllocator();
+        if (pLoaderAllocator->IsCollectible())
+        {
+            pShuffleThunkCache = ((AssemblyLoaderAllocator*)pLoaderAllocator)->GetShuffleThunkCache();
+        }
 
-    LoaderAllocator* pLoaderAllocator = pDelMT->GetLoaderAllocator();
-    if (pLoaderAllocator->IsCollectible())
+        pShuffleThunk = pShuffleThunkCache->Canonicalize((const BYTE *)&rShuffleEntryArray[0]);
+    }
+    else
     {
-        pShuffleThunkCache = ((AssemblyLoaderAllocator*)pLoaderAllocator)->GetShuffleThunkCache();
+#if defined(TARGET_RISCV64)
+        pShuffleThunk = CreateILDelegateShuffleThunk(pMD, isInstRetBuff);
+#else
+        _ASSERTE(FALSE);
+        return NULL;
+#endif // TARGET_RISCV64
     }
 
-    Stub* pShuffleThunk = pShuffleThunkCache->Canonicalize((const BYTE *)&rShuffleEntryArray[0]);
     if (!pShuffleThunk)
     {
         COMPlusThrowOM();
     }
 
-    if (!pTargetMeth->IsStatic() && pTargetMeth->HasRetBuffArg() && IsRetBuffPassedAsFirstArg())
-    {
-        if (InterlockedCompareExchangeT(&pClass->m_pInstRetBuffCallStub, pShuffleThunk, NULL ) != NULL)
-        {
-            ExecutableWriterHolder<Stub> shuffleThunkWriterHolder(pShuffleThunk, sizeof(Stub));
-            shuffleThunkWriterHolder.GetRW()->DecRef();
-            pShuffleThunk = pClass->m_pInstRetBuffCallStub;
-        }
-    }
-    else
+    // Cache the shuffle thunk
+    Stub** ppThunk = isInstRetBuff ? &pClass->m_pInstRetBuffCallStub : &pClass->m_pStaticCallStub;
+    Stub* pExistingThunk = InterlockedCompareExchangeT(ppThunk, pShuffleThunk, NULL);
+    if (pExistingThunk != NULL)
     {
-        if (InterlockedCompareExchangeT(&pClass->m_pStaticCallStub, pShuffleThunk, NULL ) != NULL)
-        {
-            ExecutableWriterHolder<Stub> shuffleThunkWriterHolder(pShuffleThunk, sizeof(Stub));
-            shuffleThunkWriterHolder.GetRW()->DecRef();
-            pShuffleThunk = pClass->m_pStaticCallStub;
-        }
+        ExecutableWriterHolder<Stub> shuffleThunkWriterHolder(pShuffleThunk, sizeof(Stub));
```

Comment:
I would expect that `ExecutableWriterHolder` is going to fail "The RX block to map as RW was not found" when called on the IL-based shuffle thunk. The Stub wrapper for IL-based shuffle thunks is not allocated in the executable memory.

Could you please check that this path works?

It may be a good idea to temporarily enable IL-based shuffle thunks unconditionally and run all tests to check for issues this one.

Reply:
Thank you, that always passed because RISC-V doesn't support W^X with crossgen2 yet (#103408). Fixed, tested with crossgen2 in SPC.dll turned off, DOTNET_EnableWriteXorExecute=1, and impromptu forcing double creation of shuffle thunks.